File: receiver_android/app/src/test/java/com/iflytek/lanmediacast/receiver/security/SasCalculatorTest.kt
package com.iflytek.lanmediacast.receiver.security

import java.io.File
import java.util.Base64
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import org.junit.Assert.assertEquals
import org.junit.Test

class SasCalculatorTest {
    @Test
    fun `reserved Phase 2 SAS matches shared fixed vector`() {
        val fixture = Json.parseToJsonElement(
            File("../../protocol/fixtures/v1/valid/sas_vector.json").readText(Charsets.UTF_8),
        ).jsonObject
        fun string(name: String) = fixture.getValue(name).jsonPrimitive.content
        fun bytes(name: String) = Base64.getUrlDecoder().decode(string(name))

        assertEquals(
            string("expectedSas"),
            SasCalculator.calculate(
                bytes("certificateSha256"),
                string("senderId"),
                bytes("senderNonce"),
                bytes("receiverNonce"),
                string("challengeId"),
            ),
        )
    }
}

File: sender_flutter/test/sas_calculator_test.dart
import 'dart:convert';
import 'dart:io';
import 'dart:typed_data';

import 'package:flutter_test/flutter_test.dart';
import 'package:lan_media_cast_sender/protocol/sas_calculator.dart';

void main() {
  test('reserved Phase 2 SAS matches the shared fixed vector', () {
    final Map<String, dynamic> fixture =
        jsonDecode(
              File(
                '../protocol/fixtures/v1/valid/sas_vector.json',
              ).readAsStringSync(),
            )
            as Map<String, dynamic>;
    Uint8List bytes(String name) => Uint8List.fromList(
      base64Url.decode(base64Url.normalize(fixture[name] as String)),
    );

    expect(
      SasCalculator.calculate(
        certificateDigest: bytes('certificateSha256'),
        senderId: fixture['senderId'] as String,
        senderNonce: bytes('senderNonce'),
        receiverNonce: bytes('receiverNonce'),
        challengeId: fixture['challengeId'] as String,
      ),
      fixture['expectedSas'],
    );
  });
}
